firedrake/slate/slac/utils.py

```diff
@@ -137,41 +136,61 @@ def visit_Symbol(self, o, *args, **kwargs):
         return SymbolWithFuncallIndexing(o.symbol, o.rank, o.offset)
 
 
-# Thanks, Miklos!
-def traverse_dags(expr_dags):
-    """Traverses a DAG and returns the unique operands associated
-    with the DAG.
+def depth_first_search(graph, node, visited, schedule):
+    """A recursive depth-first search (DFS) algorithm for
+    traversing a DAG consisting of Slate expressions.
+
+    :arg graph: A DAG whose nodes (vertices) are Slate expressions
+                with edges connected to dependent expressions.
+    :arg node: A starting vertex.
+    :arg visited: A set keeping track of visited nodes.
+    :arg schedule: A list of reverse-postordered nodes. This list is
+                   used to produce a topologically sorted list of
+                   Slate nodes.
+    """
+    if node not in visited:
+        visited.add(node)
+
+        for n in graph[node]:
+            depth_first_search(graph, n, visited, schedule)
+
+        schedule.append(node)
+
+
+def topological_sort(exprs):
+    """Topologically sorts a list of Slate expressions. The
+    expression graph is constructed by relating each Slate
+    node with a list of dependent Slate nodes.
 
-    :arg expr_dags: an iterable of Slate nodes that make up the
-                    DAG of a given Slate expression.
+    :arg exprs: A list of Slate expressions.
     """
-    seen = set()
-    container = []
+    graph = OrderedDict((expr, set(expand_dag(expr)) - {expr})
+                        for expr in exprs)
 
-    for tensor in expr_dags:
-        if tensor not in seen:
-            seen.add(tensor)
-            container.append(tensor)
+    schedule = []
+    visited = set()
+    for n in graph:
+        depth_first_search(graph, n, visited, schedule)
 
+    return schedule
+
+
+def expand_dag(expr):
+    """Expands out a Slate expressions in terms of
+    all its nodes that make up the DAG.
+
+    :arg expr: A Slate expressions.
+    """
+    seen = {expr}
+    container = [expr]
+    expanded_dag = []
     while container:
         tensor = container.pop()
-        yield tensor
+        expanded_dag.append(tensor)
+
         for operand in tensor.operands:
             if operand not in seen:
                 seen.add(operand)
                 container.append(operand)
 
-
-# This function is based on the GEM DAG reference count
-# algorithm in TSFC and adapted for Slate tensors.
-def collect_reference_count(expr_dags):
-    """Returns a mapping from operand to reference count.
-
-    :arg expr_dags: an iterable of Slate nodes that make up the
-                    DAG of a given Slate expression.
-    """
-    result = collections.Counter(expr_dags)
-    for tensor in traverse_dags(expr_dags):
-        result.update(tensor.operands)
-
-    return result
+    return expanded_dag
```

Comment:
Is it obvious why you replace the `traverse_dags` function with this renamed function that appears to nothing different except to only visit a single expression (so treats all DAGS as single-root), and isn't iterable, but instead makes and returns a list?

Reply:
I renamed it to better reflect what I'm actually using it for.

Also I felt that it wasn't necessary for me to always cast the argument to a list (`traverse_dags([foo])`) when I only ever send in one expression at a time. It made sense to me to rewrite the function to take a single argument. If you prefer the old name/implementation, then I can revert it.